### PR TITLE
otelcol.procesor.memory_limiter: new component

### LINKED ISSUE #2287: Flow: otelcol.processor.memory_limiter component
Expose the OpenTelemtry Collector `memory_limiter` processor as a Flow component called `otelcol.processor.memory_limiter`. This component is from the `otelcol` distribution.

### PR DESCRIPTION
**Pull request by tpaschalis**
Signed-off-by: Paschalis Tsilias <paschalis.tsilias@grafana.com>

#### PR Description
This PR introduces a new otelcol.processor.memory_limiter component. It can be used to guard against memory exhaustion situations on a telemetry pipeline

#### Which issue(s) this PR fixes
Closes #2287 

#### PR Checklist

- [x] CHANGELOG updated
- [X] Documentation added
- [X] Tests updated

### CODE REVIEW

File: go.sum [hunk 1]
```diff
@@ -2364,6 +2368,8 @@ github.com/shirou/gopsutil v0.0.0-20181107111621-48177ef5f880/go.mod h1:5b4v6he4
 github.com/shirou/gopsutil v2.20.9+incompatible/go.mod h1:5b4v6he4MtMOwMlS0TUMTu2PcXUg8+E1lC7eC3UO/RA=
 github.com/shirou/gopsutil v3.21.8+incompatible h1:sh0foI8tMRlCidUJR+KzqWYWxrkuuPIGiO6Vp+KXdCU=
 github.com/shirou/gopsutil v3.21.8+incompatible/go.mod h1:5b4v6he4MtMOwMlS0TUMTu2PcXUg8+E1lC7eC3UO/RA=
+github.com/shirou/gopsutil/v3 v3.22.8 h1:a4s3hXogo5mE2PfdfJIonDbstO/P+9JszdfhAHSzD9Y=
+github.com/shirou/gopsutil/v3 v3.22.8/go.mod h1:s648gW4IywYzUfE/KjXxUsqrqx/T2xO5VqOXxONeRfI=
```
Comment (tpaschalis) [inline go.sum:2372]: Go asked me for to install this to run the tests. Since we're already thinking about using gopsutil, it makes sense to add this now?

File: docs/sources/flow/reference/components/otelcol.processor.memory_limiter.md [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+Multiple `otelcol.processor.memory_limiter` components can be specified by
+giving them different labels.
+
+## Usage
+
+```river
+otelcol.processor.memory_limiter "LABEL" {
+  output {
+    check_interval = "1s"
+
```
Comment (tpaschalis) [inline docs/sources/flow/reference/components/otelcol.processor.memory_limiter.md:36]: Here, we must either define `limit` _or_ `limit_percentage`. We explain this below, but do we want to signify this here, too?

File: component/otelcol/processor/memorylimiter/memorylimiter.go [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+			args.MemoryLimitPercentage > 100 ||
+			args.MemorySpikePercentage <= 0 ||
+			args.MemorySpikePercentage > 100 {
+
+			return fmt.Errorf("limit_percentage and spike_limit_percentage must be greater than 0 and and less or equal than 100")
+		}
+		return nil
+	}
+
+	return fmt.Errorf("either limit or limit_percentage values must be set to greater than zero")
```
Comment (tpaschalis) [inline component/otelcol/processor/memorylimiter/memorylimiter.go:91]: Does this read well, or should I invert the clauses?

File: CHANGELOG.md [hunk 4]
```diff
@@ -40,6 +40,10 @@ collected logs to standard output. This can be useful for debugging. (@jcreixell
 
 - Flow: add an `otelcol.receiver.jeager` component which receives Jaeger-formatted
   traces. Data can then be forwarded to other `otelcol` components. (@rfratto)
+
+- Flow: add an `otelcol.processor.memory_limiter` component which periodically
+  check memory usage and drops data or forces a GC if the defined limits are
+  exceeded. (@tpaschalis)
```
Comment (karengermond) [inline CHANGELOG.md:46]:
```suggestion
- Flow: add an `otelcol.processor.memory_limiter` component which periodically
  checks memory usage and drops data or forces a GC if the defined limits are
  exceeded. (@tpaschalis)
```

File: docs/sources/flow/reference/components/otelcol.processor.memory_limiter.md [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+aliases:
+- /docs/agent/latest/flow/reference/components/otelcol.processor.memory_limiter
+title: otelcol.processor.memory_limiter
+---
+
+# otelcol.processor.memory_limiter
+
+`otelcol.processor.memory_limiter` is used to prevent out of memory situations
+on a telemetry pipeline. It performs periodic checks of memory usage, and will
+begin dropping data and forcing GC to reduce memory consumption if usage
```
Comment (karengermond) [inline docs/sources/flow/reference/components/otelcol.processor.memory_limiter.md:11]: Could you spell out GC?

File: component/otelcol/processor/memorylimiter/memorylimiter.go (hunk 3 above)
Comment (rfratto) [reply component/otelcol/processor/memorylimiter/memorylimiter.go:91]: It's a little odd but it's fine 👍

File: docs/sources/flow/reference/components/otelcol.processor.memory_limiter.md (hunk 2 above)
Comment (rfratto) [reply docs/sources/flow/reference/components/otelcol.processor.memory_limiter.md:36]: We should probably set one of them and then add comments saying the other can also be set.

File: docs/sources/flow/reference/components/otelcol.processor.memory_limiter.md [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+
+## Arguments
+
+`otelcol.processor.memory_limiter` supports the following arguments:
+
+
+Name | Type | Description | Default | Required
+---- | ---- | ----------- | ------- | --------
+`check_interval`     | `duration` | How often to check memory usage. | `0s` | yes
+`limit`              | `string`   | Maximum amount of memory targeted to be allocated by the process heap. | `0MiB` | no
```
Comment (rfratto) [inline docs/sources/flow/reference/components/otelcol.processor.memory_limiter.md:52]:
```suggestion
`limit`              | `string`   | Maximum amount of memory targeted to be allocated by the process heap. | `"0MiB"` | no
```

File: docs/sources/flow/reference/components/otelcol.processor.memory_limiter.md [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+`spike_limit`        | `string`   | Maximum spike expected between the measurements of memory usage. | `20% of limit` | no
```
Comment (rfratto) [inline docs/sources/flow/reference/components/otelcol.processor.memory_limiter.md:53]:
```suggestion
`spike_limit`        | `string`   | Maximum spike expected between the measurements of memory usage. | 20% of limit | no
```

File: docs/sources/flow/reference/components/otelcol.processor.memory_limiter.md [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+limits. The soft limits are then calculated as the hard limit minus the
+`spike_limit` or `spike_limit_percentage` values respectively. The recommended
+values for the spike limit values is recommended to be about 20% of the
+corresponding hard limit.
+
+The recommended `check_interval` value is 1 second. If the traffic through the
+component is spiky in nature, it is recommended to either decrease the interval
+or increase the spike limit to avoid going over the hard limit.
+
+The `limit` and `spike_limit` values must be larger than "1MiB".
```
Comment (rfratto) [inline docs/sources/flow/reference/components/otelcol.processor.memory_limiter.md:70]:
```suggestion
The `limit` and `spike_limit` values must be larger than 1 MiB.
```

File: go.sum (hunk 1 above)
Comment (rfratto) [reply go.sum:2372]: Yeah, that's fine IMO.

**Review by karengermond — APPROVED** [review]
LGTM!

File: component/otelcol/processor/memorylimiter/memorylimiter.go [hunk 9]
```diff
@@ -0,0 +1,120 @@
+// Package memorylimiter provides an otelcol.processor.memory_limiter component.
+package memorylimiter
```
Comment (rfratto) [inline component/otelcol/processor/memorylimiter/memorylimiter.go:2]: Don't forget to include this in github.com/grafana/agent/component/all so it gets defined :)

File: docs/sources/flow/reference/components/otelcol.processor.memory_limiter.md [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+- /docs/agent/latest/flow/reference/components/otelcol.processor.memory_limiter
+title: otelcol.processor.memory_limiter
+---
+
+# otelcol.processor.memory_limiter
+
+`otelcol.processor.memory_limiter` is used to prevent out of memory situations
+on a telemetry pipeline. It performs periodic checks of memory usage, and will
+begin dropping data and forcing GC to reduce memory consumption if usage
+exceeds the defined limits.
```
[the remark this answers is not in the document]
Comment (tpaschalis) [reply docs/sources/flow/reference/components/otelcol.processor.memory_limiter.md:12]: I'll slightly reword this to avoid 3x'memory usage' in two sentences, but will keep your phrasing.

File: docs/sources/flow/reference/components/otelcol.processor.memory_limiter.md [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+
+Multiple `otelcol.processor.memory_limiter` components can be specified by
+giving them different labels.
+
+## Usage
+
+```river
+otelcol.processor.memory_limiter "LABEL" {
+  output {
+    check_interval = "1s"
```
[the remark this answers is not in the document]
Comment (tpaschalis) [reply docs/sources/flow/reference/components/otelcol.processor.memory_limiter.md:35]: 🤦 sorry, fixed.

File: component/otelcol/processor/memorylimiter/memorylimiter.go [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+	}
+
+	if args.MemoryLimit > 0 {
+		if args.MemorySpikeLimit >= args.MemoryLimit {
+			return fmt.Errorf("spike_limit must be less than limit")
+		}
+		if args.MemorySpikeLimit == 0 {
+			args.MemorySpikeLimit = args.MemoryLimit / 5
+		}
+		return nil
```
[the remark this answers is not in the document]
Comment (tpaschalis) [reply component/otelcol/processor/memorylimiter/memorylimiter.go:78]: Yeah, that's how 'precedence' is enforced in the upstream component. If the fixed memory limits are present, then the percentages are ignored. Do we want to keep the same behavior, or be a little more sane and only allow one of those?

[the remark this answers is not in the document]
Comment (rfratto) [reply component/otelcol/processor/memorylimiter/memorylimiter.go:78]: Ah, yeah, I think the upstream behavior is a little confusing, and I'd prefer to start out with more strict rules. We can always loosen it later if there's complaints about it not matching upstream.

[the remark this answers is not in the document]
Comment (tpaschalis) [reply component/otelcol/processor/memorylimiter/memorylimiter.go:82]: Ok, added the clause that prohibits both of them to be set, and updated the relevant documentation parts.

File: docs/sources/flow/reference/components/otelcol.processor.memory_limiter.md [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+usage exceeds the defined limits, data is dropped and garbage collections
+are triggered to reduce it.
+
+The `memory_limiter` component uses both soft and hard limits, where the hard limit
+is always equal or larger than the soft limit. When memory usage goes above the
+soft limit, the processor component drops data and returns errors to the
+preceding components in the pipeline. When usage exceeds the hard
+limit, the processor forces a garbage collection in order to try and free
+memory. When usage drops below the soft limit, no data is dropped and no forced
+garbage collection is performed.
```
Comment (rfratto) [inline docs/sources/flow/reference/components/otelcol.processor.memory_limiter.md:20]:
```suggestion
memory. When usage is below the soft limit, no data is dropped and no forced
garbage collection is performed.
```

**Review by rfratto — APPROVED** [review]
:shipit:
